src/convert.rs

```diff
@@ -423,80 +423,137 @@ pub fn elf_to_tbf(
     // Adjust the protected region size to make fixed address work
     ////////////////////////////////////////////////////////////////////////////
 
-    // If a protected region size was passed, confirm the header will fit.
-    // Otherwise, use the header size as the protected region size.
-    let protected_region_size =
-        if let Some(fixed_protected_region_size) = protected_region_size_arg {
-            if fixed_protected_region_size < header_length as u32 {
-                // The header doesn't fit in the provided protected region size;
-                // throw an error.
-                return Err(io::Error::new(
-                    io::ErrorKind::InvalidInput,
-                    format!(
+    // Applications can hint a desired protected region size to elf2tab by
+    // defining a special `tbf_header` symbol. This symbol marks the intended
+    // load-address for the final TBF binary in flash. The protected region size
+    // is then determined by the difference between this section and the
+    // determined `fixed_address_flash` (the first address which the app
+    // requires to be loaded in flash).
+    let tbf_header_symbol = if let Ok(Some((symtab, sym_strtab))) = elf_file.symbol_table() {
+        // We are looking for the `tbf_header` symbol and its value. If it
+        // exists, we can use it as a hint for the protected region size.
+        symtab
+            .iter()
+            .find(|sym| {
+                let name = sym_strtab
+                    .get(sym.st_name as usize)
+                    .expect("Failed to parse symbol name");
+                name == "tbf_header"
+            })
+            .map(|tbf_header_sym| tbf_header_sym.st_value as u32)
+    } else {
+        None
+    };
+
+    // If a protected region size was passed, confirm the header will fit. Then,
+    // if this is a non-PIC app, fall back onto the `tbf_header` symbol for
+    // inferring the protected region size. If neither are given, use the header
+    // size as the protected region size.
+    let protected_region_size = if let Some(fixed_protected_region_size) = protected_region_size_arg
```

Comment:
This formatting looks weird, but it's been automatically adjusted by `rustfmt` and does not indent the if-block contents by a full 8 spaces, I kept it this way. Let me know if I should try change it.